<a id='menu'></a>
 <hr style="width:100%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:0.25"> 

 ![learning academy and data science campus logos](../images/la_dsc_logo.jpg)
  <hr style="width:100%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:0.25"> 

# Introduction to Python
## Chapter 5 – Cleaning Data
***
Follow along with the code by running cells as you encounter them
***
*Chapter Overview/ Learning Objectives*

* [Packages and Datasets](#packages)
 * Packages
 * Data
 
 
* [Copies and Views](#copies)


* [Updating Values](#updating_values)
 * Using `.loc[]`
 * Using `.str` methods
 * Changing column types
 

* [Changing Column Names](#column_names)
 * Changing column name values
 * Removing spaces
 * Lower Case


* [Missing Values](#missing)
 * What are mising values?
 * Null values when reading in data
 * Filling null values
 * Dropping null values


* [Tidy Data](#tidy)

This chapter covers the cleaning of our data; and this is important to do at a very early step after you read in your data.

We show this chapter after chapter 4, as some of the concepts are slightly trickier and we want you to have more experience of working with Python and Pandas before we introduce it.


<a id='packages'></a>
<hr style="width:100%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:1"> 

## Packages and Datasets

### Packages

As a reminder – we should always import our packages at the top of our script.


In this session we will use `pandas`, and give it the nickname `pd`.

Complete this action in the cell below.

In [ ]:
# Import Pandas into this cell



You can run the "solution" cell if you need help - or revisit chapter 2.

Practicing these basic commands helps your retention of the skills.

In [1]:
# %load ../solutions/chapter_5/importpandas.py
import pandas as pd

### Datasets
Good practice should also be to import our datasets at the top of our script too.

In this session we’ll be using:


|variable name |file name      |
|:--| :------|
|animals       |animals.csv    |
|titanic       |titanic.xlsx   |
|pun_animals   |pun_animals.csv|


These are all straight forward data imports with no additional parameters.

Use the cell below to load in these datasets.

If you are stuck use the solution cell; remember practicing simple commands is important for retaining the skills you’re learning.


In [ ]:
# Import your datasets here



In [2]:
# %load ../solutions/chapter_5/module_five_data.py
# load in the animals and titanic dataset )

import pandas as pd

animals = pd.read_csv("../data/animals.csv")
titanic = pd.read_excel("../data/titanic.xlsx")
pun_animals = pd.read_csv("../data/pun_animals.csv")


You can check your variables are loaded by using `%whos` in Jupyter. In Spyder or other IDE's they should appear in your variable explorer. 

If you struggle with this section – review Chapter 3.

In [3]:
%whos

Variable      Type         Data/Info
------------------------------------
animals       DataFrame           IncidentNumber    <...>n[5752 rows x 19 columns]
pd            module       <module 'pandas' from 'C:<...>es\\pandas\\__init__.py'>
pun_animals   DataFrame                  name animal<...>e           812         2
titanic       DataFrame          pclass  survived   <...>n[1309 rows x 14 columns]


[return to menu](#menu)

<a id='copies'></a>
<hr style="width:100%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:1"> 

## Copies and Views

It’s easy to think that the actions we’ve done so far, like filtering and copying create a brand new version of a DataFrame.

This isn’t always the case. Sometimes we create what’s known as a view.

On the top `cats` - highlighted in red is a subset or "view" of the original `cat_dog` DataFrame, where as on the bottom it’s a new unique object called `cats`.

![a](../images/view_copy.png)

Sometimes when experimenting with code it’s desirable to create another version of our DataFrame to experiment around on.
You may think this done as following:

`animals_new = animals`

However this creates a view of our DataFrame, and as we’re experimenting we may not want to have our effects filter though.
We can ensure we’re working on a fresh copy of a DataFrame by using `.copy()`

In [6]:
animals2 = animals.copy() # The correct way to create a copy of a DataFrame

Views and copies can potentially create issues when we try and make changes to DataFrames.

Here I’m changing the name of `Pico de Gato` to Salsa. Pico de gallo is a kind of salsa, “gato” is Spanish for cat. The best jokes are always the ones you have to explain.

Depending on my desired outcome I may want my changes to apply to **both** `cat_dog` and `cats`, like on the top by modifying a view, or I may just want them to apply to `cats` by modifying a copy like on the bottom.

![an image showing modifications to a copy and a view](../images/modify_view_copy.png)

You may see a ` SettingWithCopyWarning` warning. This is a warning, not an error so your code will still work; but if you see this you should think about what you want the outcome to be.

Common reasons for this include what’s known as chained assignment.

The cell below uses chained assignment to update the words `Bull` in the `animals2`  DataFrame to `Cow`.

Because this is happening in a slice - ` animals2[animals2["AnimalGroupParent"] == "Bull"]` we get this error.


In [7]:
animals2[animals2["AnimalGroupParent"] == "Bull"]["AnimalGroupParent"] = "cow"

C:\Python36\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


This suggests using `.loc[]` instead - we will use this method in the next section.

When the filter is applied to `animals2` it appears as if the action has not happened.

It has, but has not been applied to `animals2` , just the slice in the cell above.

In [8]:
animals2[animals2["AnimalGroupParent"]== "Bull"]

IncidentNumber    DateTimeOfCall  CalYear  FinYear   TypeOfIncident  \
1055      165500101  28/09/2010 19:17     2010  2010/11  Special Service   

      PumpCount  PumpHoursTotal  HourlyNominalCost(£)  IncidentNominalCost(£)  \
1055        1.0             3.0                   260                   780.0   

            FinalDescription AnimalGroupParent             PropertyType  \
1055  LARGE BULL IN DISTRESS              Bull  Other outdoor location    

     SpecialServiceTypeCategory  \
1055    Other animal assistance   

                                   SpecialServiceType Borough StnGroundName  \
1055  Animal assistance - Lift heavy livestock animal  Harrow        Harrow   

     AnimalClass  Code        London  
1055      Mammal  00AQ  Inner London

Certain methods update the original DataFrame and not a copy of it, for example `.loc[]`, `.fillna()` and `dropna()`

This is quite a technical element for an introduction course; don't worry if you don't understand it at the moment. The following approaches in this course are considered best practice.

Generally speaking

* All operations generate a copy
* If `inplace=True` is provided, it will modify in-place; only some operations support this
* An indexer that sets, e.g. .loc/.iloc/.iat/.at will set inplace.
* An indexer that gets on a single-dtyped object is almost always a view (depending on the memory layout it may not be that's why this is not reliable). This is mainly for efficiency. (the example from above is for .query; this will always return a copy as its evaluated by numexpr)
* An indexer that gets on a multiple-dtyped object is always a copy.
[Stack Overflow Source](https://stackoverflow.com/questions/23296282/what-rules-does-pandas-use-to-generate-a-view-vs-a-copy)

This [numpy tutorial](https://www.jessicayung.com/numpy-views-vs-copies-avoiding-costly-mistakes/) discusses copies and views in more detail. The rules for DataFrames are less explicit, but more information can be found [with this Dataquest tutorial]( https://www.dataquest.io/blog/settingwithcopywarning/)


[return to menu](#menu)

<a id='updating_values'></a>
<hr style="width:100%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:1"> 

## Updating Values

### Using .loc[]

We’ve already seen one action of `.loc[]` when we created our new Boolean column with the missing values.

`.loc[]` can be used to access groups of rows and columns by either using labels or a Boolean array. In fact you could do any of the selection we did in Chapter 4 this way. 

However here we’re going to use `.loc[]` to update some values.


In the cell below we’re looking at the unique values of the `AnimalGroupParent` column.

Here we can see two values `Cow` and `Bull` that relate to the same kind of animal.

In [9]:
animals["AnimalGroupParent"].unique()

array(['Dog', 'Fox', 'Horse', 'Rabbit', 'Unknown - Heavy Livestock Animal',
       'Squirrel', 'Cat', 'Bird', 'Unknown - Domestic Animal Or Pet',
       'Sheep', 'Deer', 'Unknown - Wild Animal', 'Snake', 'Lizard',
       'Hedgehog', 'Hamster', 'Lamb', 'Fish', 'Bull', 'Cow', 'Ferret',
       'Budgie', 'Unknown - Animal rescue from water - Farm animal',
       'Pigeon', 'Goat', 'Tortoise'], dtype=object)

For analysis purposes I want to change `Bull` to `Cow` so I can look at all of my bovine related incidents together.

In [10]:
# Update the values using .loc[]
animals.loc[animals["AnimalGroupParent"] == "Bull", "AnimalGroupParent"]  = "Cow"

# Filter to show my new DataFrame (notice the top row was Bull and is now Cow)
animals[animals["AnimalGroupParent"] == "Cow"].head()

IncidentNumber    DateTimeOfCall  CalYear  FinYear   TypeOfIncident  \
1055      165500101  28/09/2010 19:17     2010  2010/11  Special Service   
1104      188198101  06/11/2010 09:52     2010  2010/11  Special Service   
1138      207392101  09/12/2010 17:36     2010  2010/11  Special Service   
2215      100125121  26/08/2012 12:28     2012  2012/13  Special Service   
2257      113544121  19/09/2012 12:54     2012  2012/13  Special Service   

      PumpCount  PumpHoursTotal  HourlyNominalCost(£)  IncidentNominalCost(£)  \
1055        1.0             3.0                   260                   780.0   
1104        1.0             2.0                   260                   520.0   
1138        3.0             6.0                   260                  1560.0   
2215        1.0             1.0                   260                   260.0   
2257        1.0             1.0                   260                   260.0   

                             FinalDescription AnimalGroupParent  \
1055                   LARGE BULL IN DISTRESS               Cow   
1104   SICK COW STUCK AND UNABLE TO BE RAISED               Cow   
1138  COW STUCK IN MUD  WATER RESCUE LEVEL  2               Cow   
2215     COW IN DISTRESS IN GRAND UNION CANAL               Cow   
2257                         COW STUCK IN MUD               Cow   

                 PropertyType SpecialServiceTypeCategory  \
1055  Other outdoor location     Other animal assistance   
1104                    Barn     Other animal assistance   
1138             River/canal    Animal rescue from water   
2215             River/canal    Animal rescue from water   
2257     Animal harm outdoors    Other animal assistance   

                                   SpecialServiceType     Borough  \
1055  Animal assistance - Lift heavy livestock animal      Harrow   
1104  Animal assistance - Lift heavy livestock animal      Camden   
1138           Animal rescue from water - Farm animal    Havering   
2215           Animal rescue from water - Farm animal  Hillingdon   
2257                 Assist  trapped livestock animal    Havering   

     StnGroundName AnimalClass  Code        London  
1055        Harrow      Mammal  00AQ  Inner London  
1104  Kentish Town      Mammal  00AG  Inner London  
1138    Hornchurch      Mammal  00AR  Outer London  
2215       Ruislip      Mammal  00AS  Outer London  
2257    Wennington      Mammal  00AR  Outer London

In [11]:
# Update the values using .loc[]
animals.loc[animals["AnimalGroupParent"] == "Cow", "PumpCount"]  = 5

# Filter to show my new DataFrame (notice the top row was Bull and is now Cow)
animals[animals["AnimalGroupParent"] == "Cow"].head()

IncidentNumber    DateTimeOfCall  CalYear  FinYear   TypeOfIncident  \
1055      165500101  28/09/2010 19:17     2010  2010/11  Special Service   
1104      188198101  06/11/2010 09:52     2010  2010/11  Special Service   
1138      207392101  09/12/2010 17:36     2010  2010/11  Special Service   
2215      100125121  26/08/2012 12:28     2012  2012/13  Special Service   
2257      113544121  19/09/2012 12:54     2012  2012/13  Special Service   

      PumpCount  PumpHoursTotal  HourlyNominalCost(£)  IncidentNominalCost(£)  \
1055        5.0             3.0                   260                   780.0   
1104        5.0             2.0                   260                   520.0   
1138        5.0             6.0                   260                  1560.0   
2215        5.0             1.0                   260                   260.0   
2257        5.0             1.0                   260                   260.0   

                             FinalDescription AnimalGroupParent  \
1055                   LARGE BULL IN DISTRESS               Cow   
1104   SICK COW STUCK AND UNABLE TO BE RAISED               Cow   
1138  COW STUCK IN MUD  WATER RESCUE LEVEL  2               Cow   
2215     COW IN DISTRESS IN GRAND UNION CANAL               Cow   
2257                         COW STUCK IN MUD               Cow   

                 PropertyType SpecialServiceTypeCategory  \
1055  Other outdoor location     Other animal assistance   
1104                    Barn     Other animal assistance   
1138             River/canal    Animal rescue from water   
2215             River/canal    Animal rescue from water   
2257     Animal harm outdoors    Other animal assistance   

                                   SpecialServiceType     Borough  \
1055  Animal assistance - Lift heavy livestock animal      Harrow   
1104  Animal assistance - Lift heavy livestock animal      Camden   
1138           Animal rescue from water - Farm animal    Havering   
2215           Animal rescue from water - Farm animal  Hillingdon   
2257                 Assist  trapped livestock animal    Havering   

     StnGroundName AnimalClass  Code        London  
1055        Harrow      Mammal  00AQ  Inner London  
1104  Kentish Town      Mammal  00AG  Inner London  
1138    Hornchurch      Mammal  00AR  Outer London  
2215       Ruislip      Mammal  00AS  Outer London  
2257    Wennington      Mammal  00AR  Outer London

`.loc[]` takes two arguments.

The first is my conditional look up. In the `AnimalGroupParent` column of the `animals` DataFrame, look for the value that is “Bull”.

The second is the column I wish to update, here I wish to update/overwrite the column `AnimalGroupParent`.

After closing the `.loc[]` statement I use an `= ` and give my new value; here `Cow`.

When looking at the DataFrame our first row returned has the description “Bull in distress” – this was previously `Bull` in `AnimalGroupParent`.



<hr style="width:50%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:0.5"> 

## Exercise

Use `.loc[]` to change the value `Lamb` to `Sheep`

In [ ]:
animals.loc[animals["AnimalGroupParent"] == "Bull", "AnimalGroupParent"]  = "Cow"

In [ ]:
# Exercise



In [5]:
# %load ../solutions/chapter_5/loc_exercise.py
# Use `.loc[]` to change the value `Lamb` to `Sheep`

animals.loc[animals['AnimalGroupParent'] == "Lamb", "AnimalGroupParent"]  = "Sheep"

# Filter to show the head of my new DataFrame 
animals[animals['AnimalGroupParent'] == "Sheep"].head()


IncidentNumber    DateTimeOfCall  CalYear  FinYear   TypeOfIncident  \
105        44593091  18/03/2009 11:46     2009  2008/09  Special Service   
776        72228101  10/05/2010 23:26     2010  2010/11  Special Service   
2198       95952121  19/08/2012 17:05     2012  2012/13  Special Service   
2396      161316121  25/12/2012 22:57     2012  2012/13  Special Service   
3710       52040151  29/04/2015 15:10     2015  2015/16  Special Service   

      PumpCount  PumpHoursTotal  HourlyNominalCost(£)  IncidentNominalCost(£)  \
105         1.0             1.0                   255                   255.0   
776         2.0             3.0                   260                   780.0   
2198        1.0             1.0                   260                   260.0   
2396        1.0             1.0                   260                   260.0   
3710        1.0             2.0                   298                   596.0   

                  FinalDescription AnimalGroupParent          PropertyType  \
105         SHEEP IN RIVER,J15,WR1             Sheep          River/canal    
776                 LAMB IN STREAM             Sheep  Lake/pond/reservoir    
2198  LAMB TRAPPED IN WIRE FENCING             Sheep                 Fence   
2396  ONE SHEEP TRAPPED IN FENCING             Sheep  Animal harm outdoors   
3710  SHEEP TRAPPED INTO RIVER LEA             Sheep          River/canal    

     SpecialServiceTypeCategory  \
105    Animal rescue from water   
776    Animal rescue from water   
2198    Other animal assistance   
2396    Other animal assistance   
3710   Animal rescue from water   

                                     SpecialServiceType         Borough  \
105              Animal rescue from water - Farm animal  Waltham Forest   
776              Animal rescue from water - Farm animal         Enfield   
2198  Animal assistance involving wild animal - Othe...         Enfield   
2396                   Assist  trapped livestock animal         Enfield   
3710             Animal rescue from water - Farm animal         Enfield   

     StnGroundName AnimalClass  Code        London  
105      Chingford      Mammal  00BH  Outer London  
776      Chingford      Mammal  00AK  Outer London  
2198     Chingford      Mammal  00AK  Outer London  
2396       Enfield      Mammal  00AK  Inner London  
3710       Enfield      Mammal  00AK  Outer London

<hr style="width:50%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:0.5"> 

### Using .str methods

If your columns are of the `object` type (a.k.a strings or text values) we can manipulate these using `.str` methods.

This is really important where our data contains the same values but different capitalisation; remember that Python would treat “Cat” and “cat” as two different values.



In [6]:
# Note the values "cat" and "Cat" both exist.

pun_animals["animal_type"].unique()


array(['cat', 'Cat', 'dog', 'pig', 'tortoise', 'chicken'], dtype=object)

We may for example want to make all of our data in `pun_animals` DataFrame `animal_type` column lowercase so that Python treats them as one value.

In [10]:
# Change Animal Group Parent to lower

pun_animals["animal_type"] = pun_animals["animal_type"].str.lower()

# View the change by looking at the unique values

pun_animals["animal_type"].unique()

array(['cat', 'dog', 'pig', 'tortoise', 'chicken'], dtype=object)


There’s a lot of other Python String methods for you to explore [within the python documentation]( https://docs.python.org/2.4/lib/string-methods.html).


<hr style="width:100%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:1"> 

### Changing column types

One of the common frustrations in any programming language can be columns not being interpreted as we think they should. This was mentioned in Chapter 4 when we looked at `.dtypes`.

When we look at the `.dtypes` for `pun_animals` we can notice that all of our columns have read in as `object` or text values.

We would expect `age`, `weight` and `steps_per_day` to read in as numerical.

Python interprets these as `objects` because this preserves the original features of the column. This leaves how, and when, to update the column up to us.

In this section we’ll fix `age` and `weight`. 

`steps_per_day` will be fixed in the section on missing values.

In [11]:
pun_animals.dtypes

name             object
animal_type      object
age              object
vaccination        bool
Steps Per Day    object
weight_kg        object
dtype: object

Let’s fix the `age` column together.

The process of changing the data type of a column should be thought of as a 3-step process:

1.	Identify – What’s the problem
2.	Fix – Update or change the values to amend the issue
3.	Convert – Change the Data Type of the column.

The examples we’re fixing here are obviously simplified examples using simple data. 

These should hopefully give you the tools to work with your own, rather more complicated data later on.


As `pun_animals` is such a small DataFrame we could just print out the whole DataFrame or our affected Series to find the error.
This works when we only have 10 rows of data, but would be much harder to see with 10,000!

In [12]:
pun_animals["age"]

0      3
1      5
2      1
3      7
4     11
5      6
6      4
7     12
8     90
9    One
Name: age, dtype: object

Sometimes for "global" issues, like commas in numerics (100,000) we can use methods like `.head()` and `.tail()` - but there are other useful methods too:

A useful python string method is `.str.isalpha()` . This returns us a Boolean series - `True` if the string contains alphabet values (e.g a combination of A-Z) and `False` if it contains something else – like numeric values (e.g only the values 0-9).

Likewise `.str.isalnum()` will return `True` if the string contains alphabetic or numeric values, but `False` if there are symbols. 

In [13]:
pun_animals["age"].str.isalpha()

0    False
1    False
2    False
3    False
4    False
5    False
6    False
7    False
8    False
9     True
Name: age, dtype: bool

We can put this in a filter to just return the row that’s `alpha` and therefore “wrong”.

In [14]:
pun_animals[pun_animals["age"].str.isalpha()]

name animal_type  age  vaccination Steps Per Day weight_kg
9  Repecka     chicken  One         True           812         2

Here the input was `One` as a word – not a numeric.

We’ve now completed step 1 – Identify the issue.

Step 2 – Fix the issue can be done in a variety of different ways.

Here I’m going to use `.loc[]` but you can use `.str` methods too.


In [15]:
pun_animals.loc[pun_animals["age"] == "One", "age"] = 1

# Print the Series to see the change
pun_animals["age"]

# This will also work
# pun_animals = pun_animals["age"].str.replace( patt = "One", repl = "1") # Note here the "1" must be a string.

0     3
1     5
2     1
3     7
4    11
5     6
6     4
7    12
8    90
9     1
Name: age, dtype: object

You may notice that even though we have now “fixed” the problem – the “One” is now 1 – the dtype of the column is still `object`.

There are a few ways of converting the type. Here we’ll use `pd.to_numeric()` - but we can also use `.astype("int64")`.


In [16]:
pun_animals["age"] = pd.to_numeric(pun_animals["age"])

# Check out the dtypes change
pun_animals["age"].dtypes

dtype('int64')

Something that’s important to note here is that this conversion using `pd.to_numeric()` **only** works because we’ve already dealt with the issue in the column that was causing it to be read as an object.

If we try running this on the `weight_kg` column; we’ll get an error. This is because we’ve not fixed the problem yet.


In [17]:
# This Cell will cause an error!

pun_animals["weight_kg"] = pd.to_numeric(pun_animals["weight_kg"])

ValueError: Unable to parse string "113kg" at position 8

<hr style="width:50%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:0.5"> 

## Exercise

Change the `weight_kg` column to a `float` datatype.

Remember the 3 step process:

1. Identify the problem
2. Fix the problem
3. Convert the Datatype

Try inverting (with a ~) the filter statement using `.str.isdigit()` may help you find the errant value!


In [24]:
pun_animals[ ~(pun_animals["weight_kg"].str.isdigit() ) ]

Empty DataFrame
Columns: [name, animal_type, age, vaccination, Steps Per Day, weight_kg]
Index: []

In [23]:
pun_animals['weight_kg'] = pun_animals['weight_kg'].str.replace("kg","")

In [26]:
pun_animals["weight_kg"] = pun_animals["weight_kg"].astype("float64")

In [29]:
pun_animals["weight_kg"].dtype

dtype('float64')

In [ ]:
# %load ../solutions/chapter_5/changing_col_types_exercise.py
# Change the weight_kg column to a int datatype.

# 1 - Identify

print(pun_animals[~pun_animals["weight_kg"].str.isdigit()])

# Filtering pun animals for the inverse (~) of the .isdigit() will return the one that is not a digit 

# 2 - Fix  Change the "113kg" value to 113

pun_animals.loc[pun_animals["weight_kg"] == "113kg", "weight_kg"] = 113

# Using .str.replace() will also work

# pun_animals['weight_kg'] = pun_animals['weight_kg'].str.replace("113kg", "113")

# Just replacing the letters kg with nothing (an empty string) makes it more generalisable!

# # pun_animals['weight_kg'] = pun_animals['weight_kg'].str.replace("kg", "")

# 3 - Convert the type

pun_animals["weight_kg"] = pun_animals["weight_kg"].astype("float64")

# print out changed column dtypes

print(pun_animals["weight_kg"].dtypes)


You may get two different errors here

``` python
Can only use .str accessor with string values, which use np.object_ dtype in pandas
```
or

``` python
C:\Python36\lib\site-packages\pandas\core\ops.py:798: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = getattr(x, name)(y)```

You’ve already changed the column to the numeric data type and you’re trying to run the replace values part of your code again – therefore this is failing. Try checking the datatype of your column, or searching for the error again to see if you need to do anything else.


[return to menu](#menu)

<a id='column_names'></a>
<hr style="width:100%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:1"> 

## Changing Column Names

We’ll often want to change column names.
As we’ve seen so far Python is not particularly fussy about what our column names are; for good practice column names should:

* Avoid spaces (use underscores),
* Not start with a number,
* Remove symbols where possible,
* Use lower case letters.

Keeping a standard convention on our column names helps us as coders to remember what they are, as well as making cross compatibility with other languages easier. Some languages can be very particular about column names.

As we’ve seen earlier we access the columns using the `.columns` attribute.


In [30]:
pun_animals.columns

Index(['name', 'animal_type', 'age', 'vaccination', 'Steps Per Day',
       'weight_kg'],
      dtype='object')

<hr style="width:75%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:0.75"> 

### Renaming Columns

We can rename columns in a variety of ways.

Here we'll look at the `.rename()` method.

The parameter `columns = ` takes a dictionary with the old name as the key and the new name as the value.

The parameter and argument `inplace=True` means that this will update the DataFrame in situ; we don’t have to worry about putting `pun_animals= ` at the start.



In [31]:
pun_animals.rename(columns={"name": "pet_name"}, inplace=True)

pun_animals.head()

pet_name animal_type  age  vaccination Steps Per Day  weight_kg
0  Catalie Portman         cat    3         True          8105        3.0
1     Pico de Gato         cat    5        False           NaN        4.0
2        Chewbarka         dog    1        False           611        6.0
3      JK Meowling         cat    7         True          8161        3.0
4               K9         dog   11         True          7081       27.0

In [35]:
pun_animals["pet_name"].head()

0    Catalie Portman
1       Pico de Gato
2          Chewbarka
3        JK Meowling
4                 K9
Name: pet_name, dtype: object

Note that now if you re-run any earlier code that uses the column "`name`" it will bring you up a `key_error` warning as this name no longer exists.

<hr style="width:75%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:0.75"> 

### Removing Spaces

To remove spaces we can use the `columns.str.replace()` method.

 `.str.replace()` takes two arguments:
 
**`pat` = The pattern, or thing we want to change. **

A space is denoted by a string with a space in it - `” “`. 


**`repl` = Replace, or the new value we want to use. **

Here an underscore; again given as a string `”_”`.


We’re performing the `.str.replace()` method on the `.columns` attribute of our DataFrame – `pun_animals`

In [36]:
pun_animals.columns = pun_animals.columns.str.replace(pat=" ", repl="_")

pun_animals.columns

Index(['pet_name', 'animal_type', 'age', 'vaccination', 'Steps_Per_Day',
       'weight_kg'],
      dtype='object')

This can also be extended by using regular expressions, if you have more complicated replaces to run. Chapter 4 contains some links to regular expression guides in the filter section.

<hr style="width:75%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:0.75"> 

### Changing to lower case

We can use other `.str.` methods on our columns attribute too if needed – things like `.columns.str.lower()` will make our column headers lower case.


In [37]:
pun_animals.columns = pun_animals.columns.str.lower()

pun_animals.head()

pet_name animal_type  age  vaccination steps_per_day  weight_kg
0  Catalie Portman         cat    3         True          8105        3.0
1     Pico de Gato         cat    5        False           NaN        4.0
2        Chewbarka         dog    1        False           611        6.0
3      JK Meowling         cat    7         True          8161        3.0
4               K9         dog   11         True          7081       27.0

There are packages available that will allow you to clean your column names using a single function.


For those of you familiar with the `R` statistical software package `Janitor` there is a Python version – [pyjanitor]( https://pyjanitor.readthedocs.io/) which has a method [ .clean_names() ](https://pyjanitor.readthedocs.io/reference/janitor.functions/janitor.clean_names.html#janitor.clean_names). 

These packages are not included in Anaconda, so require installation.


[return to menu](#menu)

<a id='missing'></a>
<hr style="width:100%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:1"> 

# Missing Values


### What are mising values?

Missing (or null) values in Pandas are represented by `NaN`.

`NaN` is an acronym for Not a Number; but `pandas` uses this value to represent missing values in both numeric (`float`) and text (`object`) based columns.

When writing code we can use the value `None` to represent missing values.

If you have the `Numpy` package installed using `import numpy as np` you can use `np.nan` to represent missing values too.

In [38]:
# First import numpy as np
# Again "proper" convention says we should do this at the top of our code
# But as we're only using it for this cell, and this is a training notebook - we'll import it here

import numpy as np

# Create a Series with some missing values
# Again - we usually don't create our own Series, but just for a demonstration here

missing_series = pd.Series([np.nan, 4, 5, 5, None, 12, 18, 17] )

# Print out the missing Series

missing_series

0     NaN
1     4.0
2     5.0
3     5.0
4     NaN
5    12.0
6    18.0
7    17.0
dtype: float64

<hr style="width:75%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:0.75"> 

### Null values when reading in data


When we read in Data using the various `pd.read_` methods `pandas` will handle some of our missing data values for us.
If we look in the doc string from `pd.read_csv()` we can see under the `na_values` parameter that there are some default values that are interpreted as missing.

```{python}

By default the following values are interpreted as
    `NaN: '', '#N/A', '#N/A N/A', '#NA', '-1.#IND', '-1.#QNAN', '-NaN', '-nan', `
    '1.#IND', '1.#QNAN', 'N/A', 'NA', 'NULL', 'NaN', 'nan'`.  
```

If you look at the `pun_animals` DataFrame in Excel, you’ll see that within the column `steps_per_day`:

 ![missing values from excel](../images/missing_value.jpg)
 
`Pico de Gato`’s value was N/A
 
`Arf Vader`’s was NaN. 

You’ll notice that I’ve also used a `.` to represent missing data for `Voldetort`.

In [39]:
pun_animals

pet_name animal_type  age  vaccination steps_per_day  weight_kg
0  Catalie Portman         cat    3         True          8105        3.0
1     Pico de Gato         cat    5        False           NaN        4.0
2        Chewbarka         dog    1        False           611        6.0
3      JK Meowling         cat    7         True          8161        3.0
4               K9         dog   11         True          7081       27.0
5        Arf Vader         dog    6        False           NaN       16.0
6        Spiderpig         pig    4         True          5995       32.0
7     Andy Warhowl         dog   12         True          6920       27.0
8        Voldetort    tortoise   90         True             .      113.0
9          Repecka     chicken    1         True           812        2.0

`Pico de Gato` and `Arf Vader` have been interpreted as the missing data type “NaN”.

`Voldetort` has not - pandas does not interpret the `.` automatically as a missing data type.

If we check the `dtypes` the column has been interpreted an object (text); because of the full stop `.`

In [40]:
pun_animals["steps_per_day"].dtypes

dtype('O')

We now have two options:

The first is telling `pandas` to recognise the "." character as a null value. If we look back in the doc string at the `na_values= ` parameter we can see that we can pass `Additional strings to recognize as NA/NaN’.`

We could reimport our data like so:

In [41]:
pd.read_csv("../data/pun_animals.csv", na_values ="." )

name animal_type  age  vaccination  Steps Per Day weight_kg
0  Catalie Portman         cat    3         True         8105.0         3
1     Pico de Gato         Cat    5        False            NaN         4
2        Chewbarka         dog    1        False          611.0         6
3      JK Meowling         cat    7         True         8161.0         3
4               K9         dog   11         True         7081.0        27
5        Arf Vader         dog    6        False            NaN        16
6        Spiderpig         pig    4         True         5995.0        32
7     Andy Warhowl         dog   12         True         6920.0        27
8        Voldetort    tortoise   90         True            NaN     113kg
9          Repecka     chicken  One         True          812.0         2

Note: Here I’m **NOT** assigning the output of the `pd.read_csv()` function to a variable. 

This is on purpose, so as not to overwrite the earlier cleaning we did!

If you want to suppress the default missing values for whatever reason the parameter

`keep_default_na = False`

Also exists. If I set this as `False` the only things Pandas will treat as missing values are the ones we specify in `na_values = `.
        

<hr style="width:75%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:0.75"> 

### Fixing the missing value in place.

The second method to correct the column is to fill and replace the value. As i've already done some cleaning and processing to this data, if we read the data back in, we’d have to perform this cleaning again.

If you know that there’s non standard missing values, it’s best to sort them out when you load in your data; but they can be handled later – using methods like `.loc[]` and then converting to a numerics with `pd.to_numeric()` like we’ve seen before.

You may notice that the columns `.dtype` is now `float64` - Older versions of Pandas do not have a missing data type for integers.

Panda’s has introduced a nullable integer data type from version 1.0.0. You can find more information about that version [in this link]( https://pandas.pydata.org/pandas-docs/stable/user_guide/integer_na.html) as the majority of our learners use an older versions of Pandas we don’t cover it in this course.

In [42]:
# Correct the issue - and set "." to None

pun_animals.loc[pun_animals["steps_per_day"] == ".", "steps_per_day"] = None

# Convert column into numbers

pun_animals["steps_per_day"] = pd.to_numeric(pun_animals["steps_per_day"])

# Check dytpes to see if it's worked

pun_animals["steps_per_day"].dtypes

dtype('float64')

As before – if you run this cell twice you’ll get an error message!

<hr style="width:75%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:0.75"> 

### Filling null values


#### Filling with a value.

We can now handle our missing values. 

There’s a wide variety of approaches you may wish to take here, you can:
* Fill your missing values with a static number
* Fill with a summary statistic (see Chapter 6)
* Drop those rows from your dataframe.
* Use more advanced packages like SciPy which give functions like [`.interpolate()` ]( https://docs.scipy.org/doc/scipy/reference/tutorial/interpolate.html) 


What you do; and how you do it should always be led by the data you have and what the best, most appropriate decision for the analysis of the data is. 

There are also editing and imputation course on the learning hub that may help your decisions. These are theory based; but can give you starting points to find python methods.

Below we're going to fill with a value -  there are multiple ways to do this within Python.

We’re going to use the `.fillna()` method.

In [43]:
# create a copy so we're not doing this on our original data

pun_animals_fill = pun_animals.copy()


# Create a value to fill my missing values with by rounding the mean of the steps per day column to the nearest whole number

missing_puns = round(number=pun_animals_fill["steps_per_day"].mean(), ndigits=0 ) # in later Python versions this may be digits = 

# Fill missing values
pun_animals_fill["steps_per_day"].fillna(value = missing_puns,
                                         inplace=True) 

# View the DataFrame

pun_animals_fill

pet_name animal_type  age  vaccination  steps_per_day  weight_kg
0  Catalie Portman         cat    3         True         8105.0        3.0
1     Pico de Gato         cat    5        False         5384.0        4.0
2        Chewbarka         dog    1        False          611.0        6.0
3      JK Meowling         cat    7         True         8161.0        3.0
4               K9         dog   11         True         7081.0       27.0
5        Arf Vader         dog    6        False         5384.0       16.0
6        Spiderpig         pig    4         True         5995.0       32.0
7     Andy Warhowl         dog   12         True         6920.0       27.0
8        Voldetort    tortoise   90         True         5384.0      113.0
9          Repecka     chicken    1         True          812.0        2.0

Here I’m passing two arguments to `fillna()`.

`value= ` is the value I wish to fill with. This can be a static value, or in this case I’m filling with the mean of the column, rounded to the nearest whole number using `round()`

We’ll look at summary statistics in the next session. 

Note the column type is still a float; but I could convert it to an int if I wished – now I have no missing values.

`inplace=True` fills the data frame in place rather than returning a new value. 

Note that this has given the tortoise 5384 steps and although we’re not discussing the statistical implications here it’s good to highlight that this can skew your data!

<hr style="width:75%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:0.75"> 

### Dropping null values

Sometimes we’ll want to drop missing values.

Again as with all of the other examples, how you handle missing values should be first thought of in an analytical fashion, then enacted in python. Appropriate methods will vary depending on your data.

We can drop missing values easily with `dropna()`.

In [44]:
# Make another copy of our pun_animals DataFrame

pun_animals_drop = pun_animals.copy()

# Drop missing rows

pun_animals_drop.dropna(axis=0, how="any", inplace= True, subset = "steps_per_day")

# Show DataFrame
pun_animals_drop

pet_name animal_type  age  vaccination  steps_per_day  weight_kg
0  Catalie Portman         cat    3         True         8105.0        3.0
2        Chewbarka         dog    1        False          611.0        6.0
3      JK Meowling         cat    7         True         8161.0        3.0
4               K9         dog   11         True         7081.0       27.0
6        Spiderpig         pig    4         True         5995.0       32.0
7     Andy Warhowl         dog   12         True         6920.0       27.0
9          Repecka     chicken    1         True          812.0        2.0

`axis=0 ` will drop rows with null values.

`how=“any” ` will drop a row if any value is null. We could use “all” to drop the row if all values are missing.

Note we’ve lost `Pico De Gato`, `Voldetort` and `Arf Vader`

Notice again here that the index has not been reset. We can do this with the `.reset_index()` method we’ve used earlier in the course.

``` python

pun_animals_drop.reset_index(drop=True, inplace=True)

```

Another commonly used parameter is `subset=[column_names]`, often we dont need to remove all of a row because one column has missing data and we can use this to help.

[return to menu](#menu)

<a id='tidy'></a>
<hr style="width:100%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:1"> 

## Tidy Data

It’s estimated that Data Scientists spend 60% to 80% of their time on cleaning and preparing data. We’ve just scraped the surface of what you can do to clean your data, and we’ve used relatively simple data for these examples.

As you work with Python you will find new scenarios and will be able to use the tools you’ve gained in these chapters to find solutions to your problems.

It’s impossible to have a section of the course about tidying data without talking about tidy data.
[The tidy data paper]( http://vita.had.co.nz/papers/tidy-data.pdf) by Hadley Wickham; written for the R programming language tries to provide a standard-ish way of organising data within a dataset.

We recommend reading the paper; as the concepts explained within are applicable across languages.

An explanation of applying the practices of tidy data in Python can be found in this tutorial by [Jean-Nicholas Hould]( https://www.jeannicholashould.com/tidy-data-in-python.html).

[return to menu](#menu)

<a id='END'></a>
<hr style="width:100%;height:4px;border-width:0;color:gray;background-color:#003d59; opacity:1"> 

## End of Chapter 5

In this chapter we’ve explored:
* The differences between a new DataFrame and a copy of a DataFrame
* Updating Values in our DataFrame using `.loc[]`, string methods
* Changing column data types
* Changing column names
* Cleaning column names
* A basic introduction to missing values
* An exploration of the concepts of Tidy Data.


You have completed chapter 5 of the Introduction to Python course. Please move on to chapter 6.

[return to menu](#menu)